### PR TITLE
chore(rfcs): Remap language compile-time type checking v1

### LINKED ISSUE #4862: Check Remap script result value type at compile-time
The Remap language returns its last expression as a value when executed. This value can be a string, bool, integer, etc.

When using Remap in the context of a so-called "condition" within Vector, such as:

```toml
[transforms.reduce]
	type = "reduce"
	ends_when.type = "remap"
	ends_when.source = "<source code here>"
```

The condition expects to receive a boolean, which determines if reduced logs are emitted.

Currently, any non-boolean value in a remap-conditional is considered to be "false". For example:

```toml
ends_when.source = 'upcase("foo")' # is interpreted by the conditional logic as "false"
```

## Goal

The goal of this RFC is to make the above example a compile-time error. The user is forced to explicitly return a boolean value in a conditional, for example by using the `to_bool` function:

```toml
ends_when.source = 'to_bool(upcase("foo"), default = false)'
```

This should not apply to the **remap transform**, as its return value is ignored (its expressions are used to manipulate the event itself, not to return a final script value). The remap transform scripts should thus be allowed to return _any_ value.

### PR DESCRIPTION
**Pull request by JeanMertz**
This RFC spun out of a discussion in https://github.com/timberio/vector/pull/4743.

It moves one more failure condition from runtime to compile/boot-time.

👀 [Rendered version](https://github.com/timberio/vector/blob/ed88e3ff4ffc161955e521a40730c4eed228ccf5/rfcs/2020-11-02-remap-strict-return-value-types.md)

closes #4862

### CODE REVIEW

File: rfcs/2020-11-02-remap-strict-return-value-types.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+We define `ResolveKind` as follows:
+
+```rust
+pub enum ResolveKind {
+    /// The expression can resolve to any value at runtime.
+    Any,
+
+    /// The expressions resolves to one of the defined [`ValueKind`]s.
+    OneOf(Vec<ValueKind>),
```
Comment (JeanMertz) [inline rfcs/2020-11-02-remap-strict-return-value-types.md:112]: I forgot to expand on what `ValueKind` means. I'll write this up later, but it's basically and enum of the kinds of values the `Value` enum can contain (string, integer, float, etc).

It should also be noted that this only allows for checking "primitive" values, as in, you can check that the value kind is a string, integer, etc, and you can also check if it's a map or array, but you can't define what value kinds are included in the map or array.

In other words, you can't define the expectation of nested value kinds, only the top-level value kind is considered.

File: rfcs/2020-11-02-remap-strict-return-value-types.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+"transient expressions" return the values they resolve to based on whatever
+expressions they themselves hold:
+
+```rust
+impl Expression for IfStatement {
+    fn resolves_to(&self) -> ResolveKind {
+        let true_resolves = self.false_expression.resolves_to();
+        let false_resolves = self.false_expression.resolves_to();
+
+        // return the combined set of true and false resolves.
```
Comment (JeanMertz) [inline rfcs/2020-11-02-remap-strict-return-value-types.md:142]: "combined set" here means that if either one returns `Any` or `Maybe`, then this expression will also return any or maybe. Other than that, the two arrays of value kinds are merged together.

**Comment by FungusHumungus** [comment]
Nice. Yes I like where this is heading!

I think one thing that this hasn't nailed down fully is that initially any input fields can have any type, or may not even exist at all. In order to use a path with any function, this would require the users to check the types of the fields beforehand.

So for example a condition could not be just:

```
contains(.foo, "bar")
```

since .foo may be a number.

We would require the user to coerce .foo to a string first:

```
.foo = to_string(.foo)
contains(.foo, "bar")
```

Some types may not always be coercible, integers for example, so this falls back to the decision of error handling or providing defaults.

An alternative could be to have an extra field in the configuration that specifies the fields and their expected types - similar to a functions parameter list. Before running Remap, Vector could validate these fields and not run the script if they fail. This could get complex with nested maps though, so would need careful thought.

**Comment by JeanMertz** [comment]
> I think one thing that this hasn't nailed down fully is that initially any input fields can have any type, or may not even exist at all. In order to use a path with any function, this would require the users to check the types of the fields beforehand.

I will make this more explicit, but it is important to note that this RFC describes how we'll use the **result value of the program**, _not_ the result value of each individual function call (or expression).

This is also why `ResolveType` has an `Any` variant, because you rightfully noticed that the `Path` expression will by-necessity always return `Any`. It is through the use of intermediate functions that the scope of the returned values types is reduced from `Any` to one or more specific `ValueKind` variants.

That is to say, your example:

```
contains(.foo, "bar")
```

Is a perfectly valid program, and `contains` specifies that it will always return a `Boolean`, _however_, it is still a _fallible_ function, and it will return an error if `.foo` is not a string (this is the same behavior as we have right now).

The scope of the RFC makes this clear:

> The proposal does not consider fallible vs infallible expressions, those are left for a future RFC.

If you want more fine-grained control over fallible functions, I have another RFC for you 😄 #4840

I would personally caution against having extremely strict type-checking of fields, at least for now, but even if others think differently, I would propose to have that discussion in a relevant issue or RFC, since this one does not propose making any changes to that concept, it only concerns itself with the expected (non-error) outcome of a program as a whole.

File: rfcs/2020-11-02-remap-strict-return-value-types.md [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+the `true_expression` or the `false_expression`.
+
+Because expressions are initialized at compile-time, we are able to have such
+"transient expressions" return the values they resolve to based on whatever
+expressions they themselves hold:
+
+```rust
+impl Expression for IfStatement {
+    fn resolves_to(&self) -> ResolveKind {
+        let true_resolves = self.false_expression.resolves_to();
```
Comment (jszwedko) [inline rfcs/2020-11-02-remap-strict-return-value-types.md:157]: Should be `self.true_expression.resolves_to()`?

**Review by jszwedko — COMMENTED** [review]
In general, this looks good to me!

I was expecting to see some mention of tagging fields and variables with their kind though. Is that unnecessary?

For example, given the program:

```
$foo = parse_json(.message)
$foo.some_field
```

Would this fail to load as we are not sure of `$foo.some_field`'s type? The user could then rectify this with:

```
$foo = parse_json(.message)
to_bool($foo.some_field)
```

Another program that should also compile:

```
$foo = parse_json(.message)
$bar = to_bool($foo.some_field)
$bar
```

A program that should **not** compile:

```
$foo = parse_json(.message)
$bar = to_int($foo.some_field)
$bar
```

File: rfcs/2020-11-02-remap-strict-return-value-types.md (hunk 3 above)
Comment (JeanMertz) [reply rfcs/2020-11-02-remap-strict-return-value-types.md:157]: Nice one, you caught my read-the-RFC Easter egg. If I were a betting man, I would have put my money on you in the first place, but I'm not, so instead of sharing my profits, I'll give you some internet fame to enjoy 😄 👍

**Comment by JeanMertz** [comment]
> I was expecting to see some mention of tagging fields and variables with their kind though. Is that unnecessary?

Variables are "transient" in that they would return the `resolves_to` value of the (`Literal`) expression they store.

However, as the RFC notes, in this initial version, a map is reported as just a "map", there is no way to determine the inner value types within a map through this implementation just yet.

> For example, given the program:
> 
> ```
> $foo = parse_json(.message)
> $foo.some_field
> ```
> 
> Would this fail to load as we are not sure of `$foo.some_field`'s type? The user could then rectify this with:
>
> Another program that should also compile:
> 
> ```
> $foo = parse_json(.message)
> $bar = to_bool($foo.some_field)
> $bar
> ```
>
> A program that should **not** compile:
> 
> ```
> $foo = parse_json(.message)
> $bar = to_int($foo.some_field)
> $bar
> ```

All of these examples currently don't compile because you can't access map data through variables right now. I do want to solve that, but I consider that outside of the scope of this RFC.

**Comment by jszwedko** [comment]
Ah, yes, I forgot that we can't index into variables yet. I think my questions are still relevant using the log object `.`. Updated examples:

For example, given the program:

```
.message = parse_json(.message)
.message.some_field
```

Would this fail to load as we are not sure of `$foo.some_field`'s type? The user could then rectify this with:

```
.message = parse_json(.message)
to_bool(.message.some_field)
```

Another program that should also compile:

```
.message = parse_json(.message)
$bar = to_bool(.message.some_field)
$bar
```

A program that should **not** compile:

```
.message = parse_json(.message)
$bar = to_int(.message.some_field)
$bar
```

I think this is related to @FungusHumungus's example that `contains(.foo, "bar")` should fail to compile without a prior `.foo = to_string(.foo)`. We can leave it out-of-scope of this RFC if you want, but I think it is a fairly natural extension of this strict interpretation of expressions.

**Comment by JeanMertz** [comment]
> For example, given the program:
> 
> ```
> .message = parse_json(.message)
> .message.some_field
> ```
> 
> Would this fail to load as we are not sure of `.message.some_field`'s type?

Any path-based query results in `Any`, there is no type information (currently) available for path queries.

> The user could then rectify this with:
> 
> ```
> .message = parse_json(.message)
> to_bool(.message.some_field)
> ```

Correct, this would work as expected.

> Another program that should also compile:
> 
> ```
> .message = parse_json(.message)
> $bar = to_bool(.message.some_field)
> $bar
> ```

Right, this is an interesting use-case.

Currently, line 2 results in an `Assignment` expression, which holds the target (path or variable), and the expression to be assigned to the target. Given this last bit of information, we can know at compile-time which value type this resolves to.

For line 3, we don't have anything other than the variable identifier (`bar`) stored at compile-time. At runtime, we query the value, but the value doesn't exist in the store until the assignment is executed at runtime. This means we don't have any details at compile-time about the variable itself.

Technically, on line 2, we _could_ store the _expression_, not the resolved _value_ in some kind of store at compile-time. Then, at runtime, when the assignment is called, we actually resolve the expression and store the value in the regular variable store.

This way, line 3 would have access to the stored expression at compile-time, and thus can query for its `ResolveKind` value. Which would make this example work as expected.

One issue is that we don't have access to the runtime state at compile-time, so we'd need some kind of separate compile-time store to keep track of variable expressions, but that's doable. We'd probably have to update the `resolves_to` function to give it access to some kind of store similar to the runtime store which can be used to transfer state such as this between different lines.

Basically, something like this:

```rust
impl Expression for Variable {
    fn resolves_to(&self, state: &mut CompilerState) -> ResolveKind {
		let variable = state.variable(&self.ident).clone();

		variable.resolves_to(state)
    }
}
```

I'm not sure of the implications of moving variable assignment to compile-time, I can't think of any obvious issues. Compile-time parsing is not that different from runtime in the sense that it parses from left-to-right, top-to-bottom, so an example such as this:

```
$bar = "foo"
$bar
$bar = true
$bar
```

Would still work, as the "compile-time store" would get a record for the `bar` variable with a `Literal<Value::String>` expression assigned to it. Then on line two, it would fetch that `Literal` expression based on the `bar` identifier and request its `ResolveKind` value. On line 3, that expression would be overwritten with `Literal<Boolean>`, but that's okay, since line 2 has already been processed.

So yeah, great example, that requires a bit more work to get working, but is definitely doable.

> I think this is related to @FungusHumungus's example that `contains(.foo, "bar")` should fail to compile without a prior `.foo = to_string(.foo)`.

There is a relation there (type checking), but I don't consider them equal. The `contains` example is about type checking arguments of functions. We do most of this at runtime (there are some compile-time checks, but nothing related to the value kind), this is about type checking the result of expressions (and thus functions) in the global scope of the program.

I agree we can even go further with argument type checking, and we will (as long as it's helpful and doesn't become annoying to the end-user), but I would say that the plumbing done in this RFC will help with that use-case as well, but for now, the focus is on the resulting type of the program, not type checking function arguments.

**Comment by JeanMertz** [comment]
I pushed two updates, one to explicitly exclude compile-time function argument checking, but noting that what we're doing here will help to introduce such a feature later on (cc @FungusHumungus) and one to add a chapter on how we are going to support keeping track of variable type details (cc @jszwedko).

File: rfcs/2020-11-02-remap-strict-return-value-types.md [hunk 4]
```diff
@@ -212,6 +212,100 @@ program, the program is considered valid, and no boot-time error occurs, or else
...
+2 only knows of the variable identifier, and the global runtime store is not
+available at compile-time, there is no way to infer what value kind the variable
+foo will resolve to at runtime.
+
+To solve this problem we'll introduce a **compile-time store** to keep track of
+resolve kinds for targets (variables and paths):
+
+```rust
+struct CompilerState {
+    variables: HashMap<String, ResolveKind>,
```
Comment (jszwedko) [inline rfcs/2020-11-02-remap-strict-return-value-types.md:257]: Would query paths just be stored as their raw string form?

**Review by jszwedko — APPROVED** [review]
Looks good! I'll be happy to see this change implemented. I think providing as much compile-time safety as we can with the remap language will be a big advantage.

File: rfcs/2020-11-02-remap-strict-return-value-types.md (hunk 4 above)
Comment (JeanMertz) [reply rfcs/2020-11-02-remap-strict-return-value-types.md:257]: Right now, they would. But this is a bit in flux with the `Lookup` stuff being merged and reverted again. It's somewhat hacky right now (because target paths also allow coalesced path segments), but I plan on cleaning that up before 0.12 ships with both `Lookup` support and Remap itself.

There is a `Target` enum, which distinguishes between variables and paths, so I'll probably swap `String` out for that enum.

**Review by Hoverbear — APPROVED** [review]
This seems reasonable and well considered from the current situation. It makes sense from the technical standpoint.

From the overall product standpoint continue to wonder if this kind of expansion of remap is the right direction. My understanding of it originally was to replace add/remove/rename and maybe swimlanes, but I feel we're now at the point we're reimplementing many of the complexities/ideas of lua/tremorscript/js that made them undesirable to us for this role in the first place.

**Comment by binarylogic** [comment]
@Hoverbear I tend to agree with your overall concerns, but more so on https://github.com/timberio/vector/pull/4840. This RFC doesn't appear to propose any user-facing changes? I don't fully understand the low-level implementation details, but I didn't see any actual language changes other than raising errors at boot time, which I agree would improve the UX if ew can do it transparently in certain cases.

**Comment by Hoverbear** [comment]
Yeah, I absolutely think this PR is a good step.

**Comment by JeanMertz** [comment]
@binarylogic is correct, there is no user-facing change after this RFC, except for moving some runtime exceptions to compile-time and creating less confusing behaviour when using remap in conditions such as reduce's `ends_when`.

Given that there are no further objections, I'll go ahead and merge this RFC, so that I can start porting over the PoC into an actual implementation.
